### PR TITLE
Map benchmarks and tweaks to hashCodes

### PR DESCRIPTION
**Pull request by szarnekow**
This PR fixes a few inconsistencies with the hashCode implementation of Lazy and Tuple1. It better matches the impl of the corresponding equals method, now.

In addition to that, the new utility HashCodes is proposed, to avoid array boxing in hashCode computations.

It contains a MapBenchmark for common map operations and I propose additional accessors for the keys and values of maps as iterators.

### CODE REVIEW

File: vavr-benchmark/src/test/java/io/vavr/MemoryUsage.java [hunk 1]
```diff
@@ -67,6 +67,7 @@ static void storeMemoryUsages(int elementCount, Object target) {
             Tuple.of("^java\\.", "Java mutable @ "),
             Tuple.of("^fj\\.", "Functional Java persistent @ "),
             Tuple.of("^org\\.pcollections", "PCollections persistent @ "),
+            Tuple.of("^io\\.usethesource", "Steindorfer persistent @ "),
```
Comment (ruslansennov) [inline vavr-benchmark/src/test/java/io/vavr/MemoryUsage.java]: already added (see several rows below)

Comment (szarnekow) [reply vavr-benchmark/src/test/java/io/vavr/MemoryUsage.java]: Fixed, thanks.

File: vavr/src/main/java/io/vavr/collection/HashArrayMappedTrie.java [hunk 2]
```diff
@@ -433,7 +448,7 @@ public int size() {
 
         @Override
         public int hashCode() {
-            return Objects.hash(hash, value);
+            return HashCodes.hash(hash, value);
```
Comment (ruslansennov) [inline vavr/src/main/java/io/vavr/collection/HashArrayMappedTrie.java]: this method will be removed, see #2187

Comment (szarnekow) [reply vavr/src/main/java/io/vavr/collection/HashArrayMappedTrie.java]: Thank you for the heads-up. I resolved the merge conflict and updated the PR.

File: vavr/src/main/java/io/vavr/collection/Map.java [hunk 3]
```diff
@@ -361,6 +363,15 @@ default boolean isTraversableAgain() {
      */
     io.vavr.collection.Set<K> keySet();
 
+    /**
+     * Returns the keys contained in this map as an iterable.
+     *
+     * @return {@code Set} of the keys contained in this map.
```
Comment (ruslansennov) [inline vavr/src/main/java/io/vavr/collection/Map.java]: Iterator

File: vavr/src/main/java/io/vavr/collection/Map.java [hunk 4]
```diff
@@ -634,6 +645,20 @@ default int length() {
...
 
+    /**
+     * Returns the values in this map.
+     *
+     * <pre>{@code
+     * // = Iterator.of("a", "b", "c")
+     * HashMap.of(1, "a", 2, "b", 3, "c").values()
+     * }</pre>
+     *
+     * @return a new {@link Seq}
```
Comment (ruslansennov) [inline vavr/src/main/java/io/vavr/collection/Map.java]: Iterator

File: vavr/src/main/java/io/vavr/control/HashCodes.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
+package io.vavr.control;
+
+import java.util.Objects;
+
+/**
+ * Utility that allows to compute hashcodes without var-arg conversion into arrays for common
+ * types of vavr.
```
Comment (ruslansennov) [inline vavr/src/main/java/io/vavr/control/HashCodes.java]: `@author Sebastian Zarnekow`

File: vavr/src/test/java/io/vavr/OutputTester.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+import org.junit.Test;
+
+import java.io.*;
+import java.nio.charset.Charset;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+
+/**
+ * Small utility that allows to test code which write to standard error or standard out.
```
Comment (ruslansennov) [inline vavr/src/test/java/io/vavr/OutputTester.java]: `@author Sebastian Zarnekow`

Comment (szarnekow) [reply vavr/src/test/java/io/vavr/OutputTester.java]: Done

Comment (szarnekow) [reply vavr/src/test/java/io/vavr/OutputTester.java]: Thank you for the hint. Done.

File: vavr/src/main/java/io/vavr/collection/Map.java [hunk 7]
```diff
@@ -289,7 +289,7 @@ default boolean containsValue(V value) {
     default <U> Seq<U> flatMap(Function<? super Tuple2<K, V>, ? extends Iterable<? extends U>> mapper) {
         Objects.requireNonNull(mapper, "mapper is null");
         // don't remove cast, doesn't compile in Eclipse without it
-        return (Seq<U>) iterator().flatMap(mapper).toStream();
```
Comment (ruslansennov) [inline vavr/src/main/java/io/vavr/collection/Map.java]: I'm not sure that it is still important, but take into account comment about "don't remove cast..."

Comment (szarnekow) [reply vavr/src/main/java/io/vavr/collection/Map.java]: damn it, that was not my intention :)

Comment (nfekete) [reply vavr/src/main/java/io/vavr/collection/Map.java]: Or if you're using Eclipse somehow, and it compiles, it's probably safe to remove.

Comment (szarnekow) [reply vavr/src/main/java/io/vavr/collection/Map.java]: I checked 'cleanup' in idea and it removed the cast ... so for now I restored it.

**Comment by ruslansennov** [comment]
thank you

**Pull request merged**